src/System.Security.Cryptography.X509Certificates/src/Internal/Cryptography/Pal.OSX/ChainPal.cs

```diff
@@ -302,7 +314,47 @@ private void BuildAndSetProperties(Tuple<X509Certificate2, int>[] elementTuples)
             ChainStatus = rollupElement.ChainElementStatus;
         }
 
-        private static void FixupStatus(X509Certificate2 cert, ref int dwStatus)
+        private static void FixupRevocationStatus(
+            Tuple<X509Certificate2, int>[] elements,
+            X509RevocationFlag revocationFlag)
+        {
+            if (revocationFlag == X509RevocationFlag.ExcludeRoot)
+            {
+                // When requested
+                int idx = elements.Length - 1;
+                Tuple<X509Certificate2, int> element = elements[idx];
+                X509ChainStatusFlags statusFlags = (X509ChainStatusFlags)element.Item2;
+
+                // Apple will terminate the chain at the first "root" or "trustAsRoot" certificate
+                // it finds, which it refers to as "anchors". We'll consider a "trustAsRoot" cert
+                // as a root for the purposes of ExcludeRoot. So as long as the last element doesn't
+                // have PartialChain consider it the root.
+                if ((statusFlags & X509ChainStatusFlags.PartialChain) == 0)
+                {
+                    statusFlags &= ~RevocationRelevantFlags;
+                    elements[idx] = Tuple.Create(element.Item1, (int)statusFlags);
+                }
+            }
+            else if (revocationFlag == X509RevocationFlag.EndCertificateOnly)
+            {
+                // In Windows the EndCertificateOnly flag (CERT_CHAIN_REVOCATION_CHECK_END_CERT) will apply
+                // to a root if that's the only element, so we'll do the same.
+                // Start at element 1, and move to the end.
+                for (int i = 1; i < elements.Length; i++)
+                {
+                    Tuple<X509Certificate2, int> element = elements[i];
+                    X509ChainStatusFlags statusFlags = (X509ChainStatusFlags)element.Item2;
+
+                    statusFlags &= ~RevocationRelevantFlags;
+                    elements[i] = Tuple.Create(element.Item1, (int)statusFlags);
```

Comment:
So element 0 is the end certificate and the element at length-1 is the root? Then for EndCertificateOnly we remove the rev flags for all elements past the first and for ExcludeRoot we remove the rev flags for only the last element.